pilot/pkg/networking/plugin/plugin.go

```diff
@@ -46,33 +48,23 @@ type CallbackListenerInputParams struct {
 	ProxyInstances []*model.ServiceInstance
 	// ServiceInstance is the service instance colocated with the listener (applies to sidecar).
 	ServiceInstance *model.ServiceInstance
-}
-
-// CallbackListenerMutableObjects is a set of objects passed to On*Listener callbacks. Fields may be nil or empty.
-// Any lists should not be overridden, but rather only appended to.
-// Non-list fields may be mutated; however it's not recommended to do this since it can affect other plugins in the
-// chain in unpredictable ways.
-type CallbackListenerMutableObjects struct {
-	// HTTPFilters is the slice of HTTP filters for the Listener. Append to only.
-	HTTPFilters []*http_conn.HttpFilter
-	// TCPFilters is the slice of TCP filters for the Listener. Append to only.
-	TCPFilters []listener.Filter
-	// Listener is the listener being built. Any field may be modified, but changing other than appending to slices
-	// is discouraged.
-	Listener *xdsapi.Listener
+	// Service is the service colocated with the listener (applies to sidecar).
+	Service *model.Service
 }
 
 // Plugin is called during the construction of a xdsapi.Listener which may alter the Listener in any
 // way. Examples include AuthenticationPlugin that sets up mTLS authentication on the inbound Listener
 // and outbound Cluster, the mixer plugin that sets up policy checks on the inbound listener, etc.
 type Plugin interface {
-	// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
+	// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service.
+	// It returns a slice of network or HTTP filters that should be appended to the listener filter list.
 	// Can be used to add additional filters on the outbound path.
-	OnOutboundListener(in *CallbackListenerInputParams, mutable *CallbackListenerMutableObjects) error
+	OnOutboundListener(in *CallbackListenerInputParams, listener *xdsapi.Listener) ([]listener.Filter, []*http_conn.HttpFilter, error)
 
 	// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+	// It returns a slice of network or HTTP filters that should be appended to the listener filter list.
 	// Can be used to add additional filters.
-	OnInboundListener(in *CallbackListenerInputParams, mutable *CallbackListenerMutableObjects) error
+	OnInboundListener(in *CallbackListenerInputParams, listener *xdsapi.Listener) ([]listener.Filter, []*http_conn.HttpFilter, error)
```

Comment:
So with this new interface, we might need to call OnOutboundListener twice, once with nil for listerner to append filters (which caller also need to make sure to append them), and again to modify the listener itself?

(same for other plugin function)

Reply:
please take another look - per our discussion i've made changes to the way the listener is built. now, everything except filters is created and passed to plugins. a final step is to marshal the filters (needed because http filters are in serialized format that cannot easily be appended to).